pkg/image/importer/importer.go

```diff
@@ -800,3 +805,119 @@ func isDockerError(err error, code errcode.ErrorCode) bool {
 	}
 	return false
 }
+
+type retryRepository struct {
+	distribution.Repository
+	retries    int
+	maxRetries int
+}
+
+// NewRetryRepository wraps a distribution.Repository with helpers that will retry authentication failures
+// a limited number of times before giving up. This primarily avoids a DockerHub issue where public images
+// unexpectedly return a 401 error, even though the token is newly created (possibly due to clock skew).
+// Retries are only performed on idempotent operations, and every successful operation increases the retry
+// count by one up to the maximum.
+func NewRetryRepository(repo distribution.Repository, retries int) distribution.Repository {
+	return &retryRepository{
+		Repository: repo,
+		retries:    retries,
+		maxRetries: 4,
+	}
+}
+
+func (r *retryRepository) Manifests(ctx context.Context, options ...distribution.ManifestServiceOption) (distribution.ManifestService, error) {
+	s, err := r.Repository.Manifests(ctx, options...)
+	if err != nil {
+		return nil, err
+	}
+	return retryManifest{ManifestService: s, repo: r}, nil
+}
+
+func (r *retryRepository) shouldRetry(err error) bool {
+	if err == nil {
+		if r.retries < r.maxRetries {
+			r.retries++
+		}
+		return false
+	}
+	if !isDockerError(err, errcode.ErrorCodeUnauthorized) {
```

Comment:
There's one problem with that, if the image is actually private, we'll retry 4 times before failing an import :/